src/main/java/picard/illumina/IlluminaBasecallsToFastq.java

```diff
@@ -370,30 +385,42 @@ private void populateWritersFromMultiplexParams() {
      * @return FastqRecordsWriter that contains one or more FastqWriters (amount depends on read structure), all using
      * outputPrefix to determine the filename(s).
      */
-    private FastqRecordsWriter buildWriter(final File outputPrefix) {
+    private ConvertedClusterDataWriter<?> buildWriter(final File outputPrefix) {
         final File outputDir = outputPrefix.getAbsoluteFile().getParentFile();
         IOUtil.assertDirectoryIsWritable(outputDir);
         final String prefixString = outputPrefix.getName();
         final String suffixString = COMPRESS_OUTPUTS ? "fastq.gz" : "fastq";
+
+        final File[] templateFiles         = new File[readStructure.templates.length()];
+        final File[] sampleBarcodeFiles    = new File[readStructure.sampleBarcodes.length()];
+        final File[] molecularBarcodeFiles = new File[readStructure.molecularBarcode.length()];
+
         final FastqWriter[] templateWriters = new FastqWriter[readStructure.templates.length()];
         final FastqWriter[] sampleBarcodeWriters = new FastqWriter[readStructure.sampleBarcodes.length()];
         final FastqWriter[] molecularBarcodeWriters = new FastqWriter[readStructure.molecularBarcode.length()];
 
         for (int i = 0; i < templateWriters.length; ++i) {
-            final String filename = String.format("%s.%d.%s", prefixString, i + 1, suffixString);
-            templateWriters[i] = fastqWriterFactory.newWriter(new File(outputDir, filename));
+            templateFiles[i] = new File(outputDir, String.format("%s.%d.%s", prefixString, i + 1, suffixString));
         }
 
         for (int i = 0; i < sampleBarcodeWriters.length; ++i) {
-            final String filename = String.format("%s.barcode_%d.%s", prefixString, i + 1, suffixString);
-            sampleBarcodeWriters[i] = fastqWriterFactory.newWriter(new File(outputDir, filename));
+            sampleBarcodeFiles[i] = new File(outputDir, String.format("%s.barcode_%d.%s", prefixString, i + 1, suffixString));
         }
 
         for (int i = 0; i < molecularBarcodeWriters.length; ++i) {
-            final String filename = String.format("%s.index_%d.%s", prefixString, i + 1, suffixString);
-            molecularBarcodeWriters[i] = fastqWriterFactory.newWriter(new File(outputDir, filename));
+            molecularBarcodeFiles[i] = new File(outputDir, String.format("%s.index_%d.%s", prefixString, i + 1, suffixString));
+        }
+
+        if (SORT) {
```

Comment:
I agree that we should try to collapse these into a single `ClusterData` writer.